### DISCUSSION TITLE
Scroll restoration after page refresh with scrollRestoration experiment

### DISCUSSION THREAD

**Discussion by jzheng-move** [vercel/next.js#33777] [Help]
Hi everyone,

I recently enabled `scrollRestoration` experiment in our project to support SPA navigation scroll restoration, but I found that after adding this feature, the scroll position is not maintained after page refresh.

After some investigation, I found the issue is when we have `scrollRestoration: true`, `window.history.scrollRestoration` is set to `manual`. However, to enable scroll restoration after page refresh, we need `window.history.scrollRestoration` is set to `auto`. Seems like these two cases are conflicting with each other

Does anyone have the same problem or have any suggestion? 

Thanks in advance!!  😄

**Comment by przbab** [comment]
Hi,

I had the same issue recently and went down the rabbit hole to understand it. I came up with 2 options:

1. Don't use `getServerSideProps` where you want restoration to work and don't enable the experimental scroll restoration. This is how scroll restoration works perfectly on this example
https://next-blog-wordpress.vercel.app/
https://github.com/vercel/next.js/tree/canary/examples/cms-wordpress
Without the experimental scroll restoration browser's native `window.history.scrollRestoration` set to `auto` handles page refreshes like you said. When it comes to navigation scroll restoration it's working because the props for the page are cached when you are using `getStaticProps` so the default browser's implementation can handle that too. At least that's what came up with testing it. I couldn't find any documentation about this caching so there may be cases where it doesn't work. I suspect that without cache the restoration happens during the pending request and the result seems like it haven't been restored. I didn't find any way of caching `getServerSideProps` props on client side.

2. Enable experimental scroll restoration and handle the page refresh case on your own. This is what I went with since switching from `getServerSideProps` to `getStaticProps` was not an option for me. As far as I can tell the experimental implementation of scroll restoration doesn't handle refresh case by design https://github.com/vercel/next.js/commit/38bd1a024cb25923d8ea15f269a7294d073684d8. I came up with a hook that I call in `_app.js`.
```
import { useEffect } from 'react';
import { throttle } from 'throttle-debounce';

function useRefreshScrollRestoration() {
    useEffect(() => {
        const pageAccessedByReload = window.performance
            .getEntriesByType('navigation')
            .map((nav) => nav.type)
            .includes('reload');

        if (pageAccessedByReload) {
            const scrollPosition = Number.parseInt(sessionStorage.getItem('scrollPosition'), 10);
            if (scrollPosition) {
                window.scrollTo(0, scrollPosition);
            }
        }

        const handleScroll = throttle(500, () => {
            sessionStorage.setItem('scrollPosition', window.scrollY);
        });

        window.addEventListener('scroll', handleScroll);

        return () => {
            window.removeEventListener('scroll', handleScroll);
        };
    }, []);
}

export default useRefreshScrollRestoration;
```

I hope this helps you out.

**Reply by patterson412** [reply]
Excellent, I had to modify this to run only when the content is loaded on the screen, only then it worked. Thank you so much, i tried all other solutions nothing worked but this !.
